torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -39,78 +26,116 @@ void UpdateTorchValueByOnnxValueInfo(
 
-  v->setType(v_type);
+  return v_type;
 }
 
-bool IsSupportedNode(const Node* n) {
-  auto node_kind = n->kind();
+ListTypePtr TorchListTypeFromONNX(
+    const onnx::TypeProto_Sequence& onnx_sequence_type) {
+  c10::optional<at::ScalarType> scalar_type;
+  if (onnx_sequence_type.has_elem_type()) {
+    auto onnx_seq_elem_type = onnx_sequence_type.elem_type();
+    if (onnx_seq_elem_type.has_tensor_type()) {
+      auto onnx_tensor_type = onnx_seq_elem_type.tensor_type();
+      auto v_tensor_type = TorchTensorTypeFromONNX(onnx_tensor_type);
+      auto v_type = ListType::create(v_tensor_type);
+      return v_type;
+    }
+  }
+  return nullptr;
+}
 
-  if (!node_kind.is_onnx()) {
-    // node kind is not ONNX, skipped.
-    return false;
+void UpdateTorchValueByOnnxValueInfo(
+    Value* v,
+    const onnx::ValueInfoProto& p_info) {
+  if (!p_info.has_type()) {
+    return;
   }
 
-  if (node_kind == ::c10::onnx::SequenceAt ||
-      node_kind == ::c10::onnx::SplitToSequence ||
-      node_kind == ::c10::onnx::SequenceConstruct ||
-      node_kind == ::c10::onnx::SequenceEmpty ||
-      node_kind == ::c10::onnx::SequenceInsert ||
-      node_kind == ::c10::onnx::ConcatFromSequence ||
-      node_kind == ::c10::onnx::SequenceErase) {
-    // TODO: ONNX unable to do shape inference for these ops.
-    return false;
+  auto p_type = p_info.type();
+  if (p_type.has_tensor_type()) {
+    auto torch_tensor_type = TorchTensorTypeFromONNX(p_type.tensor_type());
+    if (torch_tensor_type) {
+      v->setType(torch_tensor_type);
+    }
+  } else if (p_type.has_sequence_type()) {
+    auto torch_list_type = TorchListTypeFromONNX(p_type.sequence_type());
+    if (torch_list_type) {
+      v->setType(torch_list_type);
+    }
   }
+}
+
+bool IsSupportedNode(const Node* n) {
+  auto node_kind = n->kind();
 
-  if (node_kind == ::c10::onnx::ConstantOfShape) {
-    // && n->input()->node()->kind() == ::c10::prim::ListConstruct
-    // TODO: ONNX shape inference segfault.
+  if (!node_kind.is_onnx()) {
+    // node kind is not ONNX, skipped.
     return false;
   }
 
-  if (node_kind == ::c10::onnx::Loop || node_kind == ::c10::onnx::If) {
-    // TODO: Support Loop & If shape inference by propagating input shape to
-    // block input.
+  // Skip when block size is zero. This is when the node is first created, doesn't have subblocks attached yet.
+  // Run shape inference for these nodes when the subgraph has already completed shape inferencing.
+  if ((node_kind == ::c10::onnx::Loop || node_kind == ::c10::onnx::If) && n->blocks().size() == 0) {
     return false;
   }
 
   return true;
 }
 
-} // namespace
-
-void ONNXShapeTypeInference(Node* n, int opset_version) {
-  if (!IsSupportedNode(n)) {
-    return;
-  }
-
-  // Create a Graph containing only the single node n.
-  // This graph is later converted to ONNX to run shape inference.
-  auto n_graph = std::make_shared<Graph>();
+Node* CloneNodeForInference(Node* n, std::shared_ptr<Graph> n_graph) {
```

Comment:
The method name is not immediately evocative. Why `..ForInference`?

Reply:
Changed to `CloneNodeToGraph`